File: src/phase2.jl
function phase2(lp::LinearProgram, init_basis::BasisState, rule::PivotRule,
                logger::SimplexLogger;
                maxit::Int = 1000, tol::Float64 = 1e-10) :: SimplexResult

    A, b, c = lp.A, lp.b, lp.c
    n       = size(A, 2)

    basic    = copy(init_basis.basic)
    nonbasic = copy(init_basis.nonbasic)

    x         = zeros(n)
    x[basic]  = A[:, basic] \ b

    log_phase(logger, 2)

    for it in 1:maxit
        B     = A[:, basic]
        N     = A[:, nonbasic]
        BF    = lu(B)
        xb    = BF \ b
        BinvN = BF \ N

        cb = c[basic]
        cr = c[nonbasic] .- BinvN' * cb
        z  = dot(cb, xb)

        log_iteration(logger, it, BasisState(copy(basic), copy(nonbasic)), xb, z)

        j = entering_index(rule, cr; tol)
        if isnothing(j)
            x          .= 0.0
            x[basic]    = xb
            return SimplexResult(x, z, OPTIMAL, it)
        end

        d = BinvN[:, j]
        i = leaving_index(xb, d, BinvN; tol)

        if isnothing(i)
            ray          = zeros(n)
            ray[nonbasic[j]] = 1.0
            ray[basic]       = -d
            return SimplexResult(ray, Inf, UNBOUNDED, it)
        end

        theta = xb[i] / d[i]

        x          .= 0.0
        x[basic]    = xb .- theta .* d
        x[nonbasic[j]] = theta
        x[basic[i]]    = 0.0  # numerical cleanup for the leaving variable

        basic[i], nonbasic[j] = nonbasic[j], basic[i]
    end

    error("Phase 2 did not converge in $maxit iterations — possible cycling")
end

File: typescript/src/phase2.ts
import type { LinearProgram, BasisState, SimplexResult } from './types.js';
import type { PivotRule } from './pivot.js';
import { enteringIndex, leavingIndex } from './pivot.js';
import { luFactor, luSolve, luSolveMatrix, dot } from './linalg.js';
import { getColumns } from './linalg.js';
import type { SimplexLogger } from './logger.js';

export function phase2(
  lp: LinearProgram,
  initBasis: BasisState,
  rule: PivotRule,
  logger: SimplexLogger,
  maxit: number = 1000,
  tol: number = 1e-10,
): SimplexResult {
  const { A, b, c } = lp;
  const n = A[0]?.length ?? 0;

  const basic = initBasis.basic.slice();
  const nonbasic = initBasis.nonbasic.slice();

  logger.logPhase(2);

  for (let it = 1; it <= maxit; it++) {
    const B = getColumns(A, basic);
    const N = getColumns(A, nonbasic);
    const BF = luFactor(B);
    const xb = luSolve(BF, b);
    const BinvN = luSolveMatrix(BF, N);

    const cb = basic.map(j => c[j]);
    // cr[j] = c[nonbasic[j]] - BinvN[:,j]' * cb
    const cr = nonbasic.map((j, col) => {
      let s = c[j];
      for (let row = 0; row < cb.length; row++) s -= BinvN[row][col] * cb[row];
      return s;
    });
    const z = dot(cb, xb);

    logger.logIteration(it, { basic: basic.slice(), nonbasic: nonbasic.slice() }, xb, z);

    const j = enteringIndex(rule, cr, tol);
    if (j === null) {
      const x = new Array(n).fill(0);
      for (let i = 0; i < basic.length; i++) x[basic[i]] = xb[i];
      return { x, z, status: 'OPTIMAL', iterations: it };
    }

    const d = BinvN.map(row => row[j]);
    const i = leavingIndex(xb, d, BinvN, tol);

    if (i === null) {
      const ray = new Array(n).fill(0);
      ray[nonbasic[j]] = 1.0;
      for (let r = 0; r < basic.length; r++) ray[basic[r]] = -d[r];
      return { x: ray, z: Infinity, status: 'UNBOUNDED', iterations: it };
    }

    const theta = xb[i] / d[i];
    const x = new Array(n).fill(0);
    for (let r = 0; r < basic.length; r++) x[basic[r]] = xb[r] - theta * d[r];
    x[nonbasic[j]] = theta;
    x[basic[i]] = 0; // numerical cleanup

    // swap entering and leaving
    const tmp = basic[i];
    basic[i] = nonbasic[j];
    nonbasic[j] = tmp;
  }

  throw new Error(`Phase 2 did not converge in ${maxit} iterations — possible cycling`);
}
